### ISSUE TITLE
Implement `union` types

### ISSUE THREAD

**Issue by ra0x3** [FuelLabs/fuel-indexer#988]
- There are various `enum` types used on the explorer - where these types contain some type of data
  - Example: `TransactinStatus` and `Transaction`
- We currently handle those by creating a wrapper entity with optional fields, and adding each enum type to the entity
- Example:
```graphql
type Transaction {
  create: CreateTransaction
  mint: MintTransaction
  script: ScriptTransaction
}
```
- Per Brandon, we should also create a type that flattens out these enums into a single type
  - Where for each `enum` variant, we add those variant fields to a single `type` 
    - If the particular `enum` variant does not have a given field, that field is `None`

So if I have a particular enum type:

```rust 
enum Foo {
  Bar {
    a, 
    b
  },
  Zoo { 
    c,
    d
  }
```
- The "flattened/rolled-up" version of this would look like:

```graphql

enum FooType {
  Bar
  Zoo
}

type Foo {
  a: UInt8
  b: UInt8
  c: UInt8
  d: UInt8
  type: FooType!
}
```

- Where when `Foo` is a `Bar` you get:
```graphql
Foo { 
  a: 1, 
  b: 1,
  c: None,
  d: None,
  type: FooType::Bar
}
```
- So on and so forth.
- This gives users the ability to query a single table in order to find entities vs having to query 2 separate tables (in this case)

**Comment by ra0x3** [comment]
How this is actually implemented in #992 

- A `union` is a type that consists of members, each of these members being a pre-defined `type` in the GraphQL schema
- The set of the union's members must all be `@virtual` types, or all _not_ `@virtual` types (no mixing and matching)
- The `type` that is derived from the union's members has a few properties:
  - The derived type is a `struct` (no different than how we derive a GraphQL `type`)
  - The derived type's fields are the set of fields from _all_ union members
    - Why? So that we can flatten enum-like types into a single table. So when I want to query for transactions, I (as the user) don't have to care about whether its a create/script, I can just query one generic transactions table. (Same with transaction status, inputs,  and outputs)
  - Why didn't we make these derived `union` types `enums`?
    - Because if we made them `enum` types, we still be storing each enum variant in its own table because each variant has its own set of fields

This was the suggested approach because fuel-core does something similar. [Here is an example of a union-like `Transaction` type.](https://github.com/FuelLabs/fuel-core/blob/141669577c73bd0a6a67a6e6ab6796385b7f5b4a/crates/client/assets/schema.sdl#L682)

**Comment by ra0x3** [comment]
@Voxelot 
- So https://github.com/FuelLabs/fuel-indexer/pull/992 is the basic implementation of union types
- Then https://github.com/FuelLabs/fuel-indexer/pull/993 actually integrates these union types into the explorer
- This edge case is something I actually ran into [and left a comment about](https://github.com/FuelLabs/fuel-indexer/pull/993/files#diff-26a796f4bcbe3e9520eb220f1fa81392e021da49cbda7569e582d2e6fb93c4d4R476)
  - The compiler will scream at you for multiple fields with different types, but we can make the error a little more user friendly definitely
    - I'll make that change in PR 993
